### PR TITLE
ENH: Added ignore_nan parameter to numpy.average.

### LINKED ISSUE #21375: ENH: np.nanmean with weights
### Proposed new feature or change:

My motivation comes from this PR in sklearn https://github.com/scikit-learn/scikit-learn/pull/23183/files#diff-b62de604fe871c9d2dcbfc799daf93b5bcebef1bf08341a0a7fc7562ba20b337R148, where I wrote:

```
Wrapper for np.average, with np.nan values being ignored from the average
    This is similar to np.nanmean, but allowing to pass weights as in np.average
```

The ideas would be to add this functionality to one of the two functions, i.e:

a) add weights to np.nanmean: https://numpy.org/doc/stable/reference/generated/numpy.nanmean.html
b) add option to ignore nan's for np.average : https://numpy.org/doc/stable/reference/generated/numpy.average.html

I'm aware of https://numpy.org/doc/stable/reference/generated/numpy.ma.average.html#numpy.ma.average but to be honest, I think it'd still be useful to implicitly indicate the mask with np.nan in the value, rather than having to create a masked array.

I'm happy to (at least try) do the PR myself, but would like some feedback on the idea / scope / api, and whether I missed some already existing method.

### PR DESCRIPTION
**Pull request by awestm**
Added ignore_nan parameter to numpy.average. Resolves https://github.com/numpy/numpy/issues/21375

### CODE REVIEW

**Comment by seberg** [comment]
FWIW, I agree with mhvk (and I am sure all or most other devs).  The only way that I might agree adding this option to a non-NaN named functions would be with a (probably via a short NEP) deciding on the intention of doing it for most functions.
And at that point it would have to align with what SciPy does with its `nan_policy` probably.  But even then: I am not actually sure we can add it to most functions in a very straight forward way, so it is like a bigger project (not to say it cannot be done).

File: numpy/lib/_function_base_impl.py [hunk 1]
```diff
@@ -434,6 +435,12 @@ def average(a, axis=None, weights=None, returned=False, *,
         the result will broadcast correctly against the original `a`.
         *Note:* `keepdims` will not work with instances of `numpy.matrix`
         or other classes whose methods do not support `keepdims`.
+    ignore_nan : bool, optional
+        Default is `False`. If `True`, nan values are ignored
+        from either the input array `a`, or both the `a` and `weights`
+        if weights are provided. In the case where weights are
+        provided, an nan in either `a` or `weights` will result
```
Comment (marctorsoc) [inline numpy/lib/_function_base_impl.py:442]:
```suggestion
        provided, a nan in either `a` or `weights` will result
```

File: numpy/lib/_function_base_impl.py [hunk 2]
```diff
@@ -538,9 +548,19 @@ def average(a, axis=None, weights=None, returned=False, *,
                     "Length of weights not compatible with specified axis.")
 
             # setup wgt to broadcast along axis
-            wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
+            if ignore_nan:
+                wgt = np.broadcast_to(wgt, a.shape)
```
Comment (marctorsoc) [inline numpy/lib/_function_base_impl.py:552]: I don't fully understand this part. Can you add a comment? Why when ignoring nans we can just BC `wgt` to `a`'s shape?

File: numpy/lib/_function_base_impl.py [hunk 3]
```diff
@@ -538,9 +548,19 @@ def average(a, axis=None, weights=None, returned=False, *,
...
+            else:
+                wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
             wgt = wgt.swapaxes(-1, axis)
 
+        if ignore_nan:
+            if not wgt.flags.writeable:
```
Comment (marctorsoc) [inline numpy/lib/_function_base_impl.py:558]: why we suddenly care about this and all the previous operations modifying `wgt` ignored it?

**Review by marctorsoc — COMMENTED** [review]
Added some comments, but the major one by the other reviewer might be the main blocker. Thanks for the initiative!

File: numpy/lib/_function_base_impl.py (hunk 2 above)
Comment (awestm) [reply numpy/lib/_function_base_impl.py:552]: My intention here was to fully broadcast wgt to match a's that way if:

```python 
a = [[1, 2, 3, np.nan], [1, 2, 3, 4]]
wgt = [1, 1, 1, 1]
wgt_broadcasted = [[1, 1, 1, 1], [1, 1, 1, 1]]
wgt_nan_removed = [[1, 1, 1, 0], [1, 1, 1, 1]]
ans = [2, 2.5]
```

Without fully broadcasting from the start I believe I would have to set `wgt=[1, 1, 1, 0]` for all the axes. I could be misunderstanding what's going on here though. 

Seems like either way I'll be closing this.

File: numpy/lib/_function_base_impl.py (hunk 3 above)
Comment (awestm) [reply numpy/lib/_function_base_impl.py:558]: BCing makes wgt read-only, which stops _copyto from replacing nans with 0 in the following step.

**Comment by awestm** [comment]
> FWIW, I agree with mhvk (and I am sure all or most other devs). The only way that I might agree adding this option to a non-NaN named functions would be with a (probably via a short NEP) deciding on the intention of doing it for most functions. And at that point it would have to align with what SciPy does with its `nan_policy` probably. But even then: I am not actually sure we can add it to most functions in a very straight forward way, so it is like a bigger project (not to say it cannot be done).

Hey @seberg, totally understand this line of thinking. The alternative to this is introducing a weights parameter into np.nanmean however I wanted to get some input before going down that route as it will also require adding a weights parameter to np.mean, which seems like a larger contribution. 

Do you think that this is the right path to go down? If you think it is I'm happy to take a stab at it.

**Comment by awestm** [comment]
> > FWIW, I agree with mhvk (and I am sure all or most other devs). The only way that I might agree adding this option to a non-NaN named functions would be with a (probably via a short NEP) deciding on the intention of doing it for most functions. And at that point it would have to align with what SciPy does with its `nan_policy` probably. But even then: I am not actually sure we can add it to most functions in a very straight forward way, so it is like a bigger project (not to say it cannot be done).
> 
> Hey @seberg, totally understand this line of thinking. The alternative to this is introducing a weights parameter into np.nanmean however I wanted to get some input before going down that route as it will also require adding a weights parameter to np.mean, which seems like a larger contribution.
> 
> Do you think that this is the right path to go down? If you think it is I'm happy to take a stab at it.

Hey @seberg, any thoughts on this?

**Comment by ngoldbaum** [comment]
I personally think adding a `weights` parameter to `nanmean` is fine. It makes the API more consistent and doesn't open the can of worms of adding NaN-handling to the main namespace.

I'm going to close this one, please feel free to open a new PR for a followup.
